pkg/instance/service.go

```diff
@@ -265,11 +266,83 @@ func (s service) FindByNameAndGroup(instance string, group string) (*model.Insta
 	return s.instanceRepository.FindByNameAndGroup(instance, group)
 }
 
-func (s service) FindInstances(groups []*models.Group, presets bool) ([]*model.Instance, error) {
+type GroupWithInstances struct {
+	Name      string
+	Hostname  string
+	Instances []*model.Instance
+}
+
+func (s service) FindInstances(user *models.User, presets bool) ([]GroupWithInstances, error) {
+	groups := s.uniqueUserGroups(user)
+
 	groupNames := make([]string, len(groups))
 	for i, group := range groups {
 		groupNames[i] = group.Name
 	}
 
-	return s.instanceRepository.FindByGroupNames(groupNames, presets)
+	instances, err := s.instanceRepository.FindByGroupNames(groupNames, presets)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.groupsWithInstances(instances), err
+}
+
+func (s service) uniqueUserGroups(user *models.User) []*models.Group {
+	groups := append(user.Groups, user.AdminGroups...)
+	return s.removeDuplicates(groups)
+}
+
+func (s service) groupsWithInstances(instances []*model.Instance) []GroupWithInstances {
+	groups := s.uniqueInstanceGroups(instances)
+	groupsWithInstances := make([]GroupWithInstances, len(groups))
+	for i, group := range groups {
+		groupsWithInstances[i].Name = group.Name
+		groupsWithInstances[i].Hostname = group.Hostname
+		groupsWithInstances[i].Instances = s.filterByGroupId(instances, func(instance *model.Instance) bool {
+			return instance.GroupName == group.Name
+		})
+	}
+	return groupsWithInstances
+}
+
+func (s service) uniqueInstanceGroups(instances []*model.Instance) []*models.Group {
+	groups := make([]*models.Group, len(instances))
+	for i, instance := range instances {
+		groups[i] = &models.Group{Name: instance.GroupName}
+	}
+	return s.removeDuplicates(groups)
```

Comment:
Do we want to group instances by group name. If that is the case I wonder if we could write an SQL query to do this work for us. 🤔 Not sure if we would need a new gorm model to represent the particular data we get back from the custom query.

If an SQL query is not an option we could explore using a set like `make(map[string]struct{})` and populate it with the group names. This would take care of removing duplicates 😄